SpringProject/src/main/java/com/gradproject2019/users/service/UserServiceImpl.java

```diff
@@ -44,44 +45,72 @@ public UserResponseDto saveUser(UserRequestDto userRequestDto) {
 
     @Override
     public UserResponseDto findUserById(Long userId, UUID token) {
-        authServiceImpl.checkTokenExists(token);
-        return getUserById(userId);
+        authServiceImpl.getTokenById(token);
+        return new UserResponseDto().from(getUserById(userId));
+    }
+
+    @Override
+    public UserResponseDto editUser(UUID token, Long userId, UserPatchRequestDto userPatchRequestDto) {
+        checkValidUpdate(token, userId, userPatchRequestDto);
 
+        userRepository.updateUser(userId, userPatchRequestDto.getFirstName(), userPatchRequestDto.getLastName(), userPatchRequestDto.getUsername(), userPatchRequestDto.getEmail(), userPatchRequestDto.getOccupation());
+        return new UserResponseDto().from(getUserById(userId));
     }
 
-    private void checkCredentialsValidity(UserRequestDto userRequestDto) {
+    private void checkValidSave(UserRequestDto userRequestDto) {
         checkCredentials(userRequestDto.getPassword(), PASSWORD_VALIDATION_PATTERN);
         checkCredentials(userRequestDto.getUsername(), USERNAME_VALIDATION_PATTERN);
         checkCredentials(userRequestDto.getEmail(), EMAIL_VALIDATION_PATTERN);
+        checkIfUsernameExists(userRequestDto.getUsername());
```

Comment:
Look into using an `UNIQUE` index here 
https://dev.mysql.com/doc/refman/8.0/en/constraint-primary-key.html
You'll need to have another migration script that adds a `UNQIUE` index to the username and email columns. https://dev.mysql.com/doc/refman/8.0/en/alter-table.html (ask for more details if the syntax isn't clear, or use MySqlWorkbench to do it graphically)

It stops multiple emails/usernames being created at the database level.

(You may not understand this yet) 
The issue with handling duplicates at the application level is that this app with have many threads communicating with a single db instance. If 2 threads are working at the same time, you could have a situation with 2 identical usernames in the db. 

When they both query the db (_before_) writing, this check will return true - there are no usernames like that. They could then _both_ write to the db at exactly, or almost the same time, which would be succesful in both cases (without `UNIQUE` indexes) 

After this any time the user/users tried to use your app they wouldn't be able to do anything because any time they did - `checkIfUsernameExists(String username)` would return false